lightly_studio/src/lightly_studio/api/routes/video_frames_media.py

```diff
@@ -82,35 +109,141 @@ def __exit__(self, exc_type: Any, exc_val: Any, exc_tb: Any) -> None:
+
+    This function implements a thread-local cache for VideoCapture objects and
+    their underlying fsspec stream. Each thread in the thread pool maintains its
+    own independent cache, allowing safe concurrent access without locking.
+
+    Args:
+        video_path: Path to the video file.
+
+    Returns:
+        Open cv2.VideoCapture object for the video.
+
+    Raises:
+        ValueError: If the video file cannot be opened.
+    """
+    if not hasattr(_thread_local, "cap_cache"):
+        _thread_local.cap_cache = OrderedDict()
+    cache: OrderedDict[str, tuple[cv2.VideoCapture, FSSpecStreamReader]] = _thread_local.cap_cache
+
+    if video_path in cache:
+        cap, stream = cache.pop(video_path)
+        if cap.isOpened():
+            cache[video_path] = (cap, stream)  # move to end
+            return cap
+        # stale entry
+        stream.close()
+
+    # open new
+    stream = FSSpecStreamReader(video_path)
+    cap = cv2.VideoCapture(cast(Any, stream), apiPreference=cv2.CAP_FFMPEG, params=())
+    if not cap.isOpened():
+        stream.close()
+        raise ValueError(f"Could not open video: {video_path}")
+
+    cache[video_path] = (cap, stream)
+    # enforce cache size
+    while len(cache) > _CAP_CACHE_SIZE:
+        _, (old_cap, old_stream) = cache.popitem(last=False)
+        old_cap.release()
+        old_stream.close()
+
+    return cap
+
+
+def _process_video_frame(
+    video_path: str,
+    frame_number: int,
+    rotation_deg: int,
+    compressed: bool,
+) -> tuple[npt.NDArray[np.uint8], str]:
+    """Process a video frame (CPU-intensive work, runs in thread pool).
+
+    This function extracts a single frame from a video file, applies any necessary
+    transformations (rotation), and encodes it as PNG or JPEG. It uses a cached
+    VideoCapture object to avoid reopening the video file for each frame request,
+    which is especially beneficial for cloud-stored videos.
+
+    Args:
+        video_path: Path to the video file.
+        frame_number: Frame number to extract.
+        rotation_deg: Rotation in degrees.
+        compressed: Whether to encode as JPEG with lower quality.
+
+    Returns:
+        Tuple of (encoded_buffer, media_type).
+
+    Raises:
+        ValueError: If frame cannot be processed.
+    """
+    cap = _get_cached_capture(video_path)
+
+    # Seek to the correct frame and read it
+    cap.set(cv2.CAP_PROP_POS_FRAMES, frame_number)
+    ret, frame = cap.read()
+    # Do not release/close; cached for reuse
+
+    if not ret:
+        raise ValueError(f"No frame at index {frame_number}")
+
+    # Apply counter-rotation if needed
+    rotate_code = ROTATION_MAP[rotation_deg]
+    if rotate_code is not None:
+        frame = cv2.rotate(src=frame, rotateCode=rotate_code)
+
+    # Encode frame - use JPEG with lower quality when compressed, PNG otherwise
+    if compressed:
+        encode_params = [cv2.IMWRITE_JPEG_QUALITY, 60]
```

Comment:
Can we do >75? 60 might already remove some details that are required for small annotations (think of annotating a 20x20 pixel object)